### ISSUE TITLE
age: streaming header MAC calculations

### ISSUE THREAD

**Issue by DavidBuchanan314** [C2SP/C2SP#28]
TL;DR: if the header MAC were calculated slightly differently, you could write an age implementation with a guaranteed upper-bound memory usage.

### The Problem

When first parsing the headers of age file, you don't yet know the `file key`. Because you don't know the `file key`, you can't calculate the `HMAC key`.

This is unfortunate, because it means you can't meaningfully use a "streaming" API  (init/update/final) for your HMAC calculations, because you need the key before you can "init".

As an implementer, you have two options:

- Once you have the `HMAC key`, rewind and re-read the header from the input file to feed it into the HMAC. This could introduce TOCTOU issues (although I'm not sure that's really within age's threat model?), and more importantly is not possible at all if the input is a pipe etc.

- Buffer the whole header in memory as you read it, and feed that buffer into the HMAC later, once you know the `HMAC key`.

I believe most (all?) implementations opt for the latter, and in practice that works *fine*. Reasonable headers are always going to fit in memory.

But what if a file has hundreds of recipients, and you need to parse it on a system under high memory pressure? Or perhaps more realistically, what if you wanted to *guarantee* that your program never uses more than a certain amount of memory even under adversarial inputs?

### The Solution?

A simple solution could be to hash-then-HMAC. That is, rather than `HMAC(key, header)`, do `HMAC(key, hash(header))`. It seems like SHA256 would be the most sensible choice of hash.

This way, you can feed the hash while you parse the headers (via a streaming init/update/final API), and do the HMAC at the end once you have the key.

This is not without [downsides](https://crypto.stackexchange.com/a/77294), but IMHO it's a reasonable trade-off, since it allows for an implementation of age to be fully streamed, without ever having to keep potentially-large buffers in memory. Any performance overhead would be utterly negligible in comparison to the cost of doing X25519/scrypt etc., and it's probably faster in practice anyway due to not having to allocate as much memory.

**Comment by DavidBuchanan314** [comment]
As a slightly contrived example, I wrote a script that generates a never-ending recipient stanza and piped it into `age`. After processing 3.5GB of input, `age` consumes over 17GB of memory and is summarily OOM-killed. It would also be possible to generate a fully valid and conforming file of finite-but-large size to achieve the same effect (of course, a ~4GB age file on its own should not raise any eyebrows).

A hypothetical fully streamed hash-then-mac implementation would not have this limitation.

![image](https://user-images.githubusercontent.com/13520633/219961071-c3f9897d-d8f5-4230-a6fa-6e1a822f5825.png)

OOM scenarios can be unpredictable - the first time I ran the script, the kernel killed my text editor. Perhaps I should file this as a bug in the age repo...

(I tried this against `rage` and rather than OOMing, it slowed to a crawl after about a megabyte - looks like there's some accidentally-quadratic time complexity in long stanza parsing, another bug for me to file...)
